### DISCUSSION TITLE
Pass a GITHUB_TOKEN from Github actions to AWS Codebuild, in order to update checks from there

### DISCUSSION THREAD

**Discussion by atirabassi-eb** [community/community#101890] [Actions] [answered]
### Select Topic Area

Question

### Body

I'm trying to integrate github actions and codebuild as automatically as possible. We're using OIDC to trigger codebuild and push code to an s3 bucket. Codebuild runs and executes tests, as expected, but when using the GITHUB_TOKEN provided by the action to update our check, we get an error:
```
{
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest"
}
```

the command we're running is this:
```
[Container] 2024/01/24 16:17:08.941915 Running command curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $GH_TOKEN" -H "X-GitHub-Api-Version: 2022-11-28" https://api.github.com/repos/REDACTED/REDACTED/statuses/$GITHUB_SHA -d '{"state":"success","target_url":"https://example.com/build/status","description":"The build succeeded!","context":"continuous-integration"}'
```

I added this to the action yaml:
```
permissions:
  id-token: write
  contents: read
  checks: write
  statuses: write
```

is it possible to use these tokens outside of github actions? should I use a github app for this purpose?
Thanks

**Comment by ghost** [comment, marked as answer]
Integrating GitHub Actions with AWS CodeBuild and attempting to use the GITHUB_TOKEN outside of GitHub Actions is a common challenge. The GITHUB_TOKEN is a short-lived token generated by GitHub Actions and is intended for use within the context of GitHub Actions workflows only. This is likely the reason you're encountering the "Bad credentials" error. The token loses its validity when used outside the Actions environment.

Here’s how you can address this issue:

1. Using a Personal Access Token (PAT): Instead of relying on GITHUB_TOKEN, you can create a Personal Access Token (PAT) in GitHub with the required scopes (repo, workflow, write:checks, etc.). This token can then be securely passed to AWS CodeBuild. Here’s how to do it:

- In your GitHub account, go to Settings → Developer settings → Personal access tokens → Generate new token.
- Select the scopes needed for your checks and updates.
- Use this token in your CodeBuild environment by setting it as an environment variable. Make sure to handle this token securely.

2. Updating the Workflow: In your GitHub Actions workflow, pass this PAT to AWS CodeBuild as an environment variable. Ensure it’s encrypted and securely handled.

3. Modify the CodeBuild Command: Update your curl command in AWS CodeBuild to use this PAT instead of the GITHUB_TOKEN:
```
curl -L -X POST -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $PAT" ...
```
Replace $PAT with the environment variable that holds your Personal Access Token.

4. Consider GitHub App: If managing a PAT is not ideal, another approach is to use a GitHub App with the necessary permissions. This method is more secure and scalable but requires additional setup. You can create a GitHub App, install it in your repository, and use its generated token in AWS CodeBuild.

In summary, the GITHUB_TOKEN is not suitable for use outside GitHub Actions. You should either use a Personal Access Token or set up a GitHub App for such purposes. Both methods require careful handling of the token to ensure security.

I hope this helps solve your integration challenge!

**Reply by atirabassi-eb** [reply]
Thank you Hovhannes! Already got started with a Github app

**Comment by MacAndersonUche** [comment]
Hi @atirabassi-eb  how did you manage to solve it? I am having` secretOrPrivateKey must be an asymmetric key when using RS256`
